autogluon/utils/tabular/ml/models/tabular_nn/tabular_nn_model.py

```diff
@@ -253,31 +247,39 @@ def train_net(self, train_dataset, test_dataset=None,
             logging.debug("initialized")
         if setup_trainer:
             # Also setup mxboard to monitor training if visualizer has been specified:
-            visualizer = self.params.get('visualizer', 'none')
+            visualizer = params.get('visualizer', 'none')
             if visualizer == 'tensorboard' or visualizer == 'mxboard':
                 try_import_mxboard()
                 from mxboard import SummaryWriter
                 self.summary_writer = SummaryWriter(logdir=self.path, flush_secs=5, verbose=False)
-            self.setup_trainer(train_dataset)
+            self.optimizer = self.setup_trainer(params=params, train_dataset=train_dataset)
         best_val_metric = -np.inf  # higher = better
         val_metric = None
         best_val_epoch = 0
-        num_epochs = self.params['num_epochs']
+        num_epochs = params['num_epochs']
         if test_dataset is not None:
             y_test = test_dataset.get_labels()
         else:
             y_test = None
 
+        if params['loss_function'] is None:
+            if self.problem_type == REGRESSION:
+                params['loss_function'] = gluon.loss.L1Loss()
+            elif self.problem_type == SOFTCLASS:
+                params['loss_function'] = gluon.loss.SoftmaxCrossEntropyLoss(sparse_label=False, from_logits=self.model.from_logits)
+            else:
+                params['loss_function'] = gluon.loss.SoftmaxCrossEntropyLoss(from_logits=self.model.from_logits)
+
```

Comment:
Is there an advantage to moving this outside of `setup_trainer`? I guess to not have to store loss_func as instance variable? 
The original philosophy was to separate `train_net` from `setup_trainer` so that `setup_trainer` could be called just once, and `train_net` could be called repeatedly to resume training on the same network (without having to setup_trainer again).  Of course moving this quick loss-function setup has little practical effect.  It is also unclear what should be the desired behavior if the user has changed `params` in between two `train_net` calls...

Reply:
Mostly it is to avoid hiding unnecessary param setting outside of the main code. loss_function is now fully defined and used in just a single function, rather than defined in one and used in another. loss_function has no dependence on setup_trainer, as I have changed it to simply return optimizer without setting any class variables.

Anything that takes virtually no time to do (such as creating the loss function) we should prioritize clarity and stability rather than speed.

I'd argue the 'unclear what would be desired behavior...' was also present before. This PR isn't aimed at answering that question.